pandas/src/period.pyx

```diff
@@ -0,0 +1,1121 @@
+                                 "because it cannot be inferred")
+
+        elif isinstance(value, compat.string_types) or lib.is_integer(value):
+            if lib.is_integer(value):
+                value = str(value)
+            value = value.upper()
+
+            dt, _, reso = parse_time_string(value, freq)
+            if freq is None:
+                try:
+                    freq = frequencies.Resolution.get_freq(reso)
+                except KeyError:
+                    raise ValueError("Invalid frequency or could not infer: %s" % reso)
+
+        elif isinstance(value, datetime):
+            dt = value
+            if freq is None:
+                raise ValueError('Must supply freq for datetime value')
+        elif isinstance(value, date):
+            dt = datetime(year=value.year, month=value.month, day=value.day)
+            if freq is None:
+                raise ValueError('Must supply freq for datetime value')
+        else:
+            msg = "Value must be Period, string, integer, or datetime"
+            raise ValueError(msg)
+
+        base, mult = _gfc(freq)
+        if mult != 1:
+            # TODO: Better error message - this is slightly confusing
+            raise ValueError('Only mult == 1 supported')
+
+        if self.ordinal is None:
+            self.ordinal = period_ordinal(dt.year, dt.month, dt.day,
+                                                dt.hour, dt.minute, dt.second, dt.microsecond, 0,
+                                                base)
+
+        self.freq = frequencies._get_freq_str(base)
+
+    def __eq__(self, other):
+        if isinstance(other, Period):
+            from pandas.tseries.frequencies import get_freq_code as _gfc
+            if other.freq != self.freq:
+                raise ValueError("Cannot compare non-conforming periods")
+            if self.ordinal == tslib.iNaT or other.ordinal == tslib.iNaT:
+                return False
+            return (self.ordinal == other.ordinal
+                    and _gfc(self.freq) == _gfc(other.freq))
+        return NotImplemented
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return hash((self.ordinal, self.freq))
+
+    def _add_delta(self, other):
+        from pandas.tseries import frequencies
+        if isinstance(other, (timedelta, np.timedelta64, offsets.Tick, Timedelta)):
+            offset = frequencies.to_offset(self.freq)
+            if isinstance(offset, offsets.Tick):
+                nanos = tslib._delta_to_nanoseconds(other)
+                offset_nanos = tslib._delta_to_nanoseconds(offset)
+
+                if nanos % offset_nanos == 0:
+                    if self.ordinal == tslib.iNaT:
+                        ordinal = self.ordinal
+                    else:
+                        ordinal = self.ordinal + (nanos // offset_nanos)
+                    return Period(ordinal=ordinal, freq=self.freq)
```

Comment:
to put on your todo list, this code in ``_add_delta`` looks a bit suspicous with all of this nested if-then (its the same as the original), but take a look in the next PR